texture_manager.c

```diff
@@ -410,78 +355,64 @@ static int last_accessed_compare(texture_2d *a, texture_2d *b) {
-            }
 
-            // If texture is already half-sized (or canvas) and it was used last frame,
-            if ((tex->scale > 1 || tex->is_canvas) && tex->frame_epoch == m_frame_epoch) {
-                // Do not reload this one (and maybe go over the memory limit)
-                continue;
-            }
+    /**
+     * Rules for Clearing Textures:
+     * 1. the texture must be loaded to be cleared, period
+     * 2. throw out all textures if clear_all is true, but respect rule 1
+     * 3. throw out failed textures, forcing them to reload if needed
+     * 4. throw out least-recently-used textures if we exceed our estimated memory limit
+     */
+    long adjusted_max_texture_bytes = manager->max_texture_bytes - manager->approx_bytes_to_load;
+    HASH_SRT(url_hash, manager->url_to_tex, last_accessed_compare);
+    texture_2d *tex = NULL;
+    texture_2d *tmp = NULL;
+    HASH_ITER(url_hash, manager->url_to_tex, tex, tmp) {
+        bool overLimit = manager->texture_bytes_used > adjusted_max_texture_bytes;
 
-            // we need to keep clearing until we can load the upcoming textures fine
-            // if we've cleared enough that we are under max texture bytes to use,
-            // then we are done
-            if (!clear_all && manager->texture_bytes_used <= adjusted_max_texture_bytes) {
-                break;
-            }
+        // if we reach a recently used image and still need memory, halfsize everything
+        if (!use_halfsized_textures && overLimit && tex->frame_epoch == m_frame_epoch) {
+            should_use_halfsized = true;
+        }
 
+        if (tex->loaded && (clear_all || tex->failed || overLimit)) {
             texture_2d *to_be_destroyed = tex;
             texture_manager_free_texture(manager, to_be_destroyed);
         }
+    }
 
-#if !defined(RELEASE)
+#if defined(TEXMAN_EXTRA_VERBOSE)
+    {
         LOG("{tex} Unloaded %d stale textures. Now: Texture count = %d. Bytes used = %d -> %d / %d", (int)(old_tex_count - manager->tex_count), (int)manager->tex_count, (int)old_bytes_used, (int)manager->texture_bytes_used, (int)adjusted_max_texture_bytes);
-#endif
 
-#if defined(TEXMAN_EXTRA_VERBOSE)
-        {
-            texture_2d *tex = NULL;
-            texture_2d *tmp = NULL;
-            HASH_ITER(url_hash, manager->url_to_tex, tex, tmp) {
-                TEXLOG("After: %s canvas=%d access=%d", tex->url, tex->is_canvas, (int)tex->last_accessed);
-            }
+        texture_2d *tex = NULL;
+        texture_2d *tmp = NULL;
+        HASH_ITER(url_hash, manager->url_to_tex, tex, tmp) {
+            TEXLOG("{tex} After: %s canvas=%d access=%d", tex->url, tex->is_canvas, (int)tex->last_accessed);
         }
-#endif
     }
+#endif
+
 }
 
 void texture_manager_reload_canvases(texture_manager *manager) {
     texture_2d *tex = NULL;
     texture_2d *tmp = NULL;
     HASH_ITER(url_hash, manager->url_to_tex, tex, tmp) {
-        if (tex->is_canvas || tex->is_text) {
+        if (tex->is_canvas) {
```

Comment:
It did at one points - our text rendering has gone through a lot of changes over the years :)

It's implemented differently per platform, see here for the android implementation where the flag is now being set:

https://github.com/gameclosure/native-android/blob/master/TeaLeaf/jni/platform/text_manager.cpp